In [1]:
import pandas as pd # for data manupulation or analysis
import numpy as np # for numeric calculation
import matplotlib.pyplot as plt # for data visualization
import seaborn as sns # for data visualization

In [2]:
from sklearn.datasets import load_breast_cancer
cancer_dataset = load_breast_cancer()
print("type of data : ",type(cancer_dataset))
print("keys for data set :  ",cancer_dataset.keys())
print("name of the features : ",cancer_dataset['feature_names'])

type of data :  <class 'sklearn.utils.Bunch'>
keys for data set :   dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])
name of the features :  ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


In [3]:
print(cancer_dataset['filename'])

/usr/local/lib/python3.6/dist-packages/sklearn/datasets/data/breast_cancer.csv


In [4]:
# create datafrmae
cancer_df = pd.DataFrame(np.c_[cancer_dataset['data'],cancer_dataset['target']],columns = np.append(cancer_dataset['feature_names'], ['target']))

In [5]:
cancer_df.tail(10)

mean radius  mean texture  ...  worst fractal dimension  target
559        11.51         23.93  ...                  0.08732     1.0
560        14.05         27.15  ...                  0.08321     1.0
561        11.20         29.37  ...                  0.05905     1.0
562        15.22         30.62  ...                  0.14090     0.0
563        20.92         25.09  ...                  0.09873     0.0
564        21.56         22.39  ...                  0.07115     0.0
565        20.13         28.25  ...                  0.06637     0.0
566        16.60         28.08  ...                  0.07820     0.0
567        20.60         29.33  ...                  0.12400     0.0
568         7.76         24.54  ...                  0.07039     1.0

[10 rows x 31 columns]

In [6]:
cancer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [35]:
# input variable
X = cancer_df.drop(['target'], axis = 1)
print(X.tail(6))

# output variable
y = cancer_df['target']
print(y.tail(5))

     mean radius  mean texture  ...  worst symmetry  worst fractal dimension
563        20.92         25.09  ...          0.2929                  0.09873
564        21.56         22.39  ...          0.2060                  0.07115
565        20.13         28.25  ...          0.2572                  0.06637
566        16.60         28.08  ...          0.2218                  0.07820
567        20.60         29.33  ...          0.4087                  0.12400
568         7.76         24.54  ...          0.2871                  0.07039

[6 rows x 30 columns]
564    0.0
565    0.0
566    0.0
567    0.0
568    1.0
Name: target, dtype: float64


In [13]:
# split dataset into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state= 5)

In [14]:
# Feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_sc = sc.fit_transform(X_train)
X_test_sc = sc.transform(X_test)

In [15]:
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

In [16]:
# XGBoost Classifier
# from xgboost import XGBClassifier
# xgb_classifier2 = XGBClassifier()
# xgb_classifier2.fit(X_train_sc, y_train)
# y_pred_xgb_sc = xgb_classifier2.predict(X_test_sc)
# accuracy_score(y_test, y_pred_xgb_sc)

0.9824561403508771

In [37]:
# Logistic Regression
from sklearn.linear_model import LogisticRegression
lr_classifier = LogisticRegression(random_state = 51, penalty = 'none',max_iter=100)
lr_classifier.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='none',
                   random_state=51, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [38]:
y_pred_lr = lr_classifier.predict(X_test)
print("Accuracy by logistic regression for all 30 parameters : ",accuracy_score(y_test, y_pred_lr))

Accuracy by logistic regression for all 30 parameters :  0.9736842105263158


In [39]:
# Confusion matrix

# cm = confusion_matrix(y_test, y_pred_lr)
# plt.title('Heatmap of Confusion Matrix', fontsize = 15)
# sns.heatmap(cm, annot = True)
# plt.show()

In [34]:
print(classification_report(y_test, y_pred_lr))

              precision    recall  f1-score   support

         0.0       0.98      0.96      0.97        48
         1.0       0.97      0.98      0.98        66

    accuracy                           0.97       114
   macro avg       0.97      0.97      0.97       114
weighted avg       0.97      0.97      0.97       114



In [44]:
cancer_df_meanone=cancer_df[["mean radius","mean texture","mean perimeter","mean area","mean smoothness", "mean compactness","mean concavity","mean concave points","mean symmetry","mean fractal dimension", "target"]]
print(cancer_df_meanone.head(5))

   mean radius  mean texture  ...  mean fractal dimension  target
0        17.99         10.38  ...                 0.07871     0.0
1        20.57         17.77  ...                 0.05667     0.0
2        19.69         21.25  ...                 0.05999     0.0
3        11.42         20.38  ...                 0.09744     0.0
4        20.29         14.34  ...                 0.05883     0.0

[5 rows x 11 columns]


In [45]:
# input variable
X_mean = cancer_df_meanone.drop(['target'], axis = 1)
print(X.tail(6))

# output variable
y_mean = cancer_df_meanone['target']
print(y.tail(5))

from sklearn.model_selection import train_test_split
X_train_mean, X_test_mean, y_train_mean, y_test_mean = train_test_split(X_mean, y_mean, test_size = 0.2, random_state= 5)

# Feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_mean_sc = sc.fit_transform(X_train_mean)
X_test_mean_sc = sc.transform(X_test_mean)

from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

# Logistic Regression
from sklearn.linear_model import LogisticRegression
lr_classifier = LogisticRegression(random_state = 51, penalty = 'none',max_iter=100)
lr_classifier.fit(X_train_mean, y_train_mean)

     mean radius  mean texture  ...  worst symmetry  worst fractal dimension
563        20.92         25.09  ...          0.2929                  0.09873
564        21.56         22.39  ...          0.2060                  0.07115
565        20.13         28.25  ...          0.2572                  0.06637
566        16.60         28.08  ...          0.2218                  0.07820
567        20.60         29.33  ...          0.4087                  0.12400
568         7.76         24.54  ...          0.2871                  0.07039

[6 rows x 30 columns]
564    0.0
565    0.0
566    0.0
567    0.0
568    1.0
Name: target, dtype: float64


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='none',
                   random_state=51, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [47]:
y_pred_lr_mean = lr_classifier.predict(X_test_mean)
print("Accuracy by logistic regression for all 30 parameters : ",accuracy_score(y_test_mean, y_pred_lr_mean))

Accuracy by logistic regression for all 30 parameters :  0.9649122807017544


##getting accuracy of 96% with logistic regression with mean data only(10 attributes)